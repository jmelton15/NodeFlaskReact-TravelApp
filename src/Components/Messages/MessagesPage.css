.MessagesPage-ConversationsCol {
    display: flex;
    justify-content: flex-end;
}

.MessagesPage-MessageCol {
    display: flex;
    justify-content: center;
}

.MessagesPage-MessageContainer {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height:90vh;
    width:80%;
    background-color: white;
    margin-top:25px;
    box-shadow: 5px 5px 30px 15px;
    border-radius: 20px;
}

.MessagesPage-ConverstaionContainer {
    height:90vh;
    width:90%;
    background-color: white;
    margin-top:25px;
    box-shadow: 5px 5px 30px 15px;
    border-radius: 20px;
    display: flex;
    flex-direction: column;
}

.MessagePage-ChatBox {
    margin-bottom: 30px;
    margin-top: 5px;
    display: flex;
    align-items: center;
    justify-content: space-around;
}

.Message.other {
    display: flex;
    align-items: center;
    justify-content: flex-start;
}
.Message.own {
    display: flex;
    align-items: center;
    justify-content: flex-end;
}

.MessagePage-StartMessageContainer {
    margin-left: 20px;
    display: flex;
    justify-content: center;
    font-size:200%;
    overflow-wrap: break-word;
    align-items: center;
    height: 3000px;
    color: grey;
}

#messageInput {
    width: 80%;
    height: 100px;
    padding:10px;
    border: 1px solid black;
}

#messageSubmitBtn {
    font-weight: bolder;
    background-color: #FD5E53;
}

#messageSubmitBtn:hover{
    background-color: #cc463c;
}

#message-wrapper {
    padding:10px;
    width:100%;
    height:100%;
    overflow-y: scroll;
}

#conversations-header {
    background-color: #dab681;
    border-radius: 10px;
    text-align: center;
    font-family: "AestheticViolet";
}


