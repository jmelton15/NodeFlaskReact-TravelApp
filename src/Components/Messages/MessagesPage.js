import "./MessagesPage.css";
import { Container, Row, Col } from 'reactstrap';
import {io} from "socket.io-client";
import { Redirect } from "react-router-dom";
import { useState,useEffect,useRef } from "react";
import Message from "./message";
import MessageForm from "./MessageForm";
import { NodeApi } from "../../APIRequests/nodeApi";
import Conversation from "../Conversations/Conversation";

const MessagesPage = ({user,token}) => { 
    const socket = useRef();
    const scrollRef = useRef();
    const [messages,setMessages] = useState(null);
    const [toUser,setToUser] = useState(null);
    const [socketMessage,setSocketMessage] = useState(null);
    
/******************************************************************************* */
    /**** WebSocket useEffect Code  ****/
    useEffect(() => {
        socket.current = io("ws://localhost:8001");
        socket.current.on("getMessage",msgData => {
            setSocketMessage({
                fromUserId:msgData.fromUserId,
                toUserId:msgData.toUserId,
                msgTxt:msgData.msgTxt
            })
        })
    },[])

    useEffect(() => {
        if(socketMessage && user.user_id === socketMessage.toUserId) {
            setMessages(prevMsg =>[...prevMsg,socketMessage])
        } 
    },[socketMessage,user.user_id]) 
 
    useEffect(() => {
        socket.current.emit("addUser",user.user_id);
        socket.current.on("getConnectedUsers",users => {
            console.log(users);
        })
    },[user,user.user_id]);
  
    useEffect(() => {
        scrollRef.current?.scrollIntoView({behavior:"smooth"});
    },[messages])

/******************************************************************************* */
    /**** Message Handling Code  ****/

    const sendMsg = (msgData) => {
        let sentMessage = NodeApi.sendMessage(msgData.message,toUser,user.user_id,user.avatar_pic_url,token)
        if(sentMessage) {
            socket.current.emit("sendMessage",{
                fromUserId:user.user_id,
                toUserId:toUser,
                msgTxt:msgData.message,
                fromUserAvatar:user.avatar_pic_url
            })
            setMessages(prevMsg =>[...prevMsg,{
                toUserId:toUser,
                fromUserId:user.user_id,
                msgTxt:msgData.message,
                fromUserAvatar:user.avatar_pic_url
            }])
        }
    }
    const getMessages = async (toUserId) => {
        let conversation = await NodeApi.getConversation(user.user_id,toUserId,token)
        setMessages(conversation);
        setToUser(toUserId);
    };

/******************************************************************************* */

    if(!token || !user) {
        return <Redirect to="/"></Redirect>
    }
    return (
        <div>
            <Container fluid>
                <Row>
                    <Col xs="8" className="MessagesPage-MessageCol">
                        <div className="MessagesPage-MessageContainer">
                            {!messages && 
                                <div className="MessagePage-StartMessageContainer"> 
                                    <blockquote>Click On A Connection To Open Up A Conversation...</blockquote>
                                </div>
                            }
                            <div id="message-wrapper">
                                {messages && messages.map((message) => {
                                return <div ref={scrollRef}>
                                            <Message message={message} currentUser={user.user_id}/>
                                        </div>
                                })}
                            </div>
                            {messages && <MessageForm sendMsg={sendMsg}/>}
                        </div>
                    </Col>
                    <Col xs="4" className="MessagesPage-ConversationsCol">
                        <div className="MessagesPage-ConverstaionContainer" onClick={() => setMessages(null)}>
                            <div id="conversations-header">
                                <h2 className="mt-1">Connections</h2>
                            </div>
                            {user.following.map((connection) => {
                                return <Conversation connection={connection} getMessages={getMessages}/>
                            })}
                        </div>
                    </Col>
                </Row>
            </Container>
        </div>
    )
}

export default MessagesPage;