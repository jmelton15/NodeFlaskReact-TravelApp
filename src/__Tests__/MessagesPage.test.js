import { render } from '@testing-library/react';
import {user} from "./TestHelperObjects";
import MessagesPage from '../Components/Messages/MessagesPage';

// simple smoke test for MessagesPage component
it("renders a MessagesPage component without crashing", function() {
    render(<MessagesPage user={user} token={true}/>);
});

// simple snapshot test for MessagesPage Component
it("should match snapshot for MessagesPage", function() {
    const {asFragment} = render (<MessagesPage user={user} token={true}/>);
    expect(asFragment()).toMatchSnapshot();
});

it("should display prompt message for no messages showing",() => {
    const {getByText} = render(<MessagesPage user={user} token={true}/>)
    const prompt = getByText("Click On A Connection To Open Up A Conversation...");
    expect(prompt).toBeInTheDocument();
})

