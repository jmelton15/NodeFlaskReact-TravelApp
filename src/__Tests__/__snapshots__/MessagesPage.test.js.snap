// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot for MessagesPage 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="MessagesPage-MessageCol col-8"
        >
          <div
            class="MessagesPage-MessageContainer"
          >
            <div
              class="MessagePage-StartMessageContainer"
            >
              <blockquote>
                Click On A Connection To Open Up A Conversation...
              </blockquote>
            </div>
            <div
              id="message-wrapper"
            />
          </div>
        </div>
        <div
          class="MessagesPage-ConversationsCol col-4"
        >
          <div
            class="MessagesPage-ConverstaionContainer"
          >
            <div
              id="conversations-header"
            >
              <h2
                class="mt-1"
              >
                Connections
              </h2>
            </div>
            <div
              class="Conversation"
            >
              <img
                alt="Avatar Pic"
                class="Conversation-AvatarPic"
                src="some pic route"
              />
              <div
                id="message-body"
              >
                <h5>
                  TESTCONN
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
